test/hotspot/gtest/nmt/test_vmatree.cpp

```diff
@@ -284,6 +284,68 @@ TEST_VM_F(NMTVMATreeTest, LowLevel) {
   }
 }
 
+TEST_VM_F(NMTVMATreeTest, SetTag) {
+  auto i = [](MemTag f) -> uint8_t { return (uint8_t)f; };
+
+  // The gc/cds case with only reserved data
+  {
+    VMATree::SummaryDiff diff;
+    Tree::RegionData rd(NCS::StackIndex(), mtNone);
+    VMATree tree;
+
+    VMATree::SummaryDiff result = tree.reserve_mapping(0, 500, rd);
+    diff.add(result);
+    EXPECT_EQ(500, diff.tag[i(mtNone)].reserve);
+
+    result = tree.reserve_mapping(500, 100, rd);
+    diff.add(result);
+    EXPECT_EQ(600, diff.tag[i(mtNone)].reserve);
+
+    result = tree.set_tag(0, 500, mtGC);
+    diff.add(result);
+    EXPECT_EQ(100, diff.tag[i(mtNone)].reserve);
+    EXPECT_EQ(500, diff.tag[i(mtGC)].reserve);
+
+    result = tree.set_tag(500, 100, mtClassShared);
+    diff.add(result);
+    EXPECT_EQ(0, diff.tag[i(mtNone)].reserve);
+    EXPECT_EQ(500, diff.tag[i(mtGC)].reserve);
+    EXPECT_EQ(100, diff.tag[i(mtClassShared)].reserve);
+  }
+
+  // Now let's add in some committed data
+  {
+    VMATree::SummaryDiff diff;
+    Tree::RegionData rd(NCS::StackIndex(), mtNone);
+    VMATree tree;
+
+    VMATree::SummaryDiff result = tree.reserve_mapping(0, 600, rd);
+    diff.add(result);
+    EXPECT_EQ(600, diff.tag[i(mtNone)].reserve);
+
+    // The committed areas
+    result = tree.commit_mapping(100, 125, rd);
+    diff.add(result);
+    result = tree.commit_mapping(550, 10, rd);
+    diff.add(result);
+    result = tree.commit_mapping(565, 10, rd);
+    diff.add(result);
+
+    // OK, set tag
+    result = tree.set_tag(0, 500, mtGC);
+    diff.add(result);
+    EXPECT_EQ(100, diff.tag[i(mtNone)].reserve);
+    EXPECT_EQ(500, diff.tag[i(mtGC)].reserve);
+    EXPECT_EQ(125, diff.tag[i(mtGC)].commit);
+
+    result = tree.set_tag(500, 100, mtClassShared);
+    diff.add(result);
+    EXPECT_EQ(0, diff.tag[i(mtNone)].reserve);
+    EXPECT_EQ(100, diff.tag[i(mtClassShared)].reserve);
+    EXPECT_EQ(20, diff.tag[i(mtClassShared)].commit);
+  }
+}
```

Comment:
I would like to see the following cases explicitly tested:
1) set a tag over a range spanning multiple existing segments with the same stack but different tags - here, I would like to see that set_tag causes the tree to coalesce these different ranges to a single range
2) set a tag over a range spanning multiple existing segments with different stacks and different tags - here, I would like to see that the original ranges stay intact, including their stack information, only the tags should now all be the same
3) set a tag within a range - here, I would like to see three ranges now

for 1, also for mismatched starts and ends where either start or end are in the middle of an existing range, which should cause the tree to split that range

Reply:
Tests added, and they all passed! I really didn't think that case 3 would work, but sometimes I write good code I guess.

Reply:
That's the good thing about tests. Gives you confidence.